### PR TITLE
Event Grid Deserializing Events Redesign

### LINKED ISSUE #13444: Design and implement EventGridConsumer
We can use the Track 1 surface as a base to work from. We will need to add support for deserializing cloud events and custom events. We should integrate with the `ObjectSerializer` abstraction as opposed to tying to any one particular JsonSerializer.

Other things to think about - what should the return type be for the Deserialize methods:
https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-method-return

For tests, we can probably port over the existing Track 1 tests and modify them to suit our needs. The tests would be unit tests as opposed to live tests since we are not communicating with the service in this type.

### LINKED ISSUE #13449: Api View for Event Grid
Now that we are moving on to the consumer side of things, it might be a good time to publish the initial API Review for Track 2 Event Grid.

### PR DESCRIPTION
**Pull request by kerri-lee**
- Removed `EventGridConsumer` class, replaced with `Parse`, `GetData<T>`, and `GetData` methods in `CloudEvent` and `EventGridEvent` types. This removes the need for users to register custom event types - now, they can simply deserialize to the type that want using `GetData<T>` or deserialize into system event types or `BinaryData` if the type is unknown using `GetData`.
- Added extension methods to `BinaryData` that will parse into a single event.
- Edited consuming event tests
- Edited live tests (due to making cloud event `Data` internal; also added additional testing for the new `BinaryData` constructor)
- Regenerated some models using updated swaggers

### CODE REVIEW

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/CloudEvent.cs [hunk 1]
```diff
@@ -24,15 +32,96 @@ public CloudEvent(string source, string type)
...
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CloudEvent"/> class.
+        /// </summary>
+        /// <param name="data"> Event data specific to the event type. </param>
+        /// <param name="source"> Identifies the context in which an event happened. The combination of id and source must be unique for each distinct event. </param>
+        /// <param name="type"> Type of event related to the originating occurrence. For example, "Contoso.Items.ItemReceived". </param>
+        /// <param name="dataContentType"> Content type of the payload. A content type different from "application/json" should be specified when sending binary data. </param>
+        public CloudEvent(BinaryData data, string source, string type, string dataContentType)
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/CloudEvent.cs:106]: `dataContentType` is optional so we should use an optional parameter.

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridEvent.cs [hunk 2]
```diff
@@ -47,5 +57,163 @@ public EventGridEvent(string subject, object data, string eventType, string data
...
+        /// </returns>
+        public object GetData()
+        {
+            if (Data != null)
+            {
+                return Data;
+            }
+
+            // Reserialize JsonElement to stream
+            return BinaryData.FromStream(SerializePayloadToStream(SerializedData, new JsonObjectSerializer()));
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridEvent.cs:208]: Use the static JsonObjectSerializer instance once we add it.

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs [hunk 3]
```diff
@@ -212,22 +216,11 @@ private async Task<Response> PublishCloudEventsInternal(IEnumerable<CloudEvent>
                     // The 'Data' property is optional for CloudEvents
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs:216]: nit: Not only is Data optional, but passing binary data into the ctor would result in this being null.

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/CloudEvent.cs [hunk 4]
```diff
@@ -38,44 +37,15 @@ public CloudEvent(string source, string type)
...
-            ExtensionAttributes = new Dictionary<string, object>();
-        }
-
-        /// <summary> Initializes a new instance of the <see cref="CloudEvent"/> class. </summary>
         /// <param name="data"> Event data specific to the event type. </param>
-        /// <param name="source"> Identifies the context in which an event happened. The combination of id and source must be unique for each distinct event. </param>
-        /// <param name="type"> Type of event related to the originating occurrence. For example, "Contoso.Items.ItemReceived". </param>
         /// <param name="dataContentType"> Content type of the payload. A content type different from "application/json" should be specified if payload is not JSON. </param>
-        public CloudEvent(object data, string source, string type, string dataContentType)
+        public CloudEvent(string source, string type, object data, string dataContentType = null)
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/CloudEvent.cs:44]: nit: use default

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs [hunk 5]
```diff
@@ -120,11 +120,18 @@ private async Task<Response> SendEventsInternal(IEnumerable<EventGridEvent> even
                     Argument.AssertNotNull(egEvent, nameof(egEvent));
 
                     JsonDocument data;
-                    using (MemoryStream stream = new MemoryStream())
+                    if (egEvent.Data is BinaryData binaryEventData)
                     {
-                        _dataSerializer.Serialize(stream, egEvent.Data, egEvent.Data.GetType(), cancellationToken);
-                        stream.Position = 0;
-                        data = JsonDocument.Parse(stream);
+                        data = JsonDocument.Parse(binaryEventData);
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs:125]: Is this using the implicit cast to ROM?

Comment (kerri-lee) [reply sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs:125]: Yeah

File: sdk/eventgrid/Azure.Messaging.EventGrid/tests/ConsumeEventTests.cs [hunk 6]
```diff
@@ -104,8 +104,14 @@ public void ConsumeEventUsingBinaryDataExtensionMethod()
...
         {
+            var customSerializer = new JsonObjectSerializer(
+                new JsonSerializerOptions()
+                {
+                    PropertyNamingPolicy = JsonNamingPolicy.CamelCase
+                });
+
             EventGridEvent egEvent = new EventGridEvent(new ContosoItemReceivedEventData(), "", "Contoso.Items.ItemReceived", "1");
-            Assert.That(() => egEvent.GetData<ContosoItemReceivedEventData>(),
+            Assert.That(() => egEvent.GetData<ContosoItemReceivedEventData>(customSerializer),
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/tests/ConsumeEventTests.cs:114]: Can we also add in a GetData calls without the serializer (both generic and non-generic)?

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs [hunk 7]
```diff
@@ -119,10 +119,20 @@ private async Task<Response> PublishEventsInternal(IEnumerable<EventGridEvent> e
                     // Individual events cannot be null
                     Argument.AssertNotNull(egEvent, nameof(egEvent));
 
-                    MemoryStream stream = new MemoryStream();
-                    _dataSerializer.Serialize(stream, egEvent.Data, egEvent.Data.GetType(), cancellationToken);
-                    stream.Position = 0;
-                    JsonDocument data = JsonDocument.Parse(stream);
+                    JsonDocument data;
+                    if (egEvent.Data is BinaryData binaryEventData)
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs:123]: Can we add a test that covers this? The generated request should have Json of BinaryData, but rather the thing that BinaryData was wrapping.

**Review by JoshLove-msft — APPROVED** [review]

File: sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs [hunk 8]
```diff
@@ -1,9 +1,15 @@
 namespace Azure.Messaging.EventGrid
 {
+    public static partial class BinaryDataExtensions
+    {
+        public static Azure.Messaging.EventGrid.CloudEvent ToCloudEvent(this Azure.Core.BinaryData binaryData) { throw null; }
+        public static Azure.Messaging.EventGrid.EventGridEvent ToEventGridEvent(this Azure.Core.BinaryData binaryData) { throw null; }
```
Comment (pakrym) [inline sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs:6]: Are we sure we want to start shipping extension methods for an exchange type like this?

Comment (JoshLove-msft) [reply sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs:6]: @KrzysztofCwalina suggested adding this

Comment (pakrym) [reply sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs:6]: VS shows extension methods from all referenced packages ignoring extensions.

image

Do we really want to make `BinaryData` an extension method dumping ground?

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs [hunk 9]
```diff
@@ -122,16 +122,18 @@ private async Task<Response> SendEventsInternal(IEnumerable<EventGridEvent> even
                     JsonDocument data;
                     if (egEvent.Data is BinaryData binaryEventData)
                     {
-                        data = JsonDocument.Parse(binaryEventData);
+                        try
+                        {
+                            data = JsonDocument.Parse(binaryEventData);
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs:127]: This somewhat highlights where a DataFormat property on BinaryData might help. We don't know whether the BinaryData is already serialized or not, so we have to do this try/catch.

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs (hunk 7 above)
Comment (JoshLove-msft) [reply sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs:123]: Turns out BinaryData can't be serialized anyway.

File: sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs (hunk 8 above)
Comment (JoshLove-msft) [reply sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs:6]: I guess something else to consider is that even when thinking strictly from service bus context, not every message will be convertible to an eg event. TryParse might be a better pattern here.

Comment (JoshLove-msft) [reply sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs:6]: I guess something else to consider is that even when thinking strictly from service bus context, not every message will be convertible to an eg event. TryParse might be a better pattern here.

**Pull request merged**